go/dpagg/count_test.go

```diff
@@ -304,22 +302,83 @@ func TestCountCheckMerge(t *testing.T) {
 				Epsilon: ln3,
 				Noise:   noise.Laplace(),
 			},
-			false,
-			false,
 			true},
 	} {
 		c1 := NewCount(tc.opt1)
 		c2 := NewCount(tc.opt2)
 
-		if tc.returnResult1 {
+		if err := checkMergeCount(c1, c2); (err != nil) != tc.wantErr {
+			t.Errorf("CheckMerge: when %v for err got %v, want %t", tc.desc, err, tc.wantErr)
+		}
+	}
+}
+
+// Tests that checkMergeCount() returns errors correctly after call to Result().
+func TestCountCheckMergeAfterResult(t *testing.T) {
+	expectedErrMessage := "checkMergeCount: %v cannot be merged with another Count instance. " +
+		"Reason: Noised result was already computed and returned."
+	for _, tc := range []struct {
+		callResult1 bool
+		callResult2 bool
+		wantErr     bool
+		errMessage  string
+	}{
+		{false, false, false, ""},
+		{true, false, true, fmt.Sprintf(expectedErrMessage, "c1")},
+		{false, true, true, fmt.Sprintf(expectedErrMessage, "c2")},
+		{true, true, true, fmt.Sprintf(expectedErrMessage, "c1")},
+	} {
+		c1 := getNoiselessCount()
+		c2 := getNoiselessCount()
+
+		if tc.callResult1 {
 			c1.Result()
 		}
-		if tc.returnResult2 {
+		if tc.callResult2 {
 			c2.Result()
 		}
 
-		if err := checkMergeCount(c1, c2); (err != nil) != tc.wantErr {
-			t.Errorf("CheckMerge: when %v for err got %v, want %t", tc.desc, err, tc.wantErr)
+		err := checkMergeCount(c1, c2)
+		if (err != nil) != tc.wantErr {
+			t.Errorf("CheckMerge: No error was returned, expected error")
+		}
+		if tc.wantErr && err.Error() != tc.errMessage {
+			t.Errorf("CheckMerge: Error messages are not equal:\n got: %v\n want: %v", err.Error(), tc.errMessage)
+		}
+	}
+}
+
+// Tests that checkMergeCount() returns errors correctly after call to GobEncode().
+func TestCountCheckMergeAfterSerialization(t *testing.T) {
+	expectedErrMessage := "checkMergeCount: %v cannot be merged with another Count instance. " +
+		"Reason: Count object has been already serialized."
+	for _, tc := range []struct {
+		callEncode1 bool
+		callEncode2 bool
+		wantErr     bool
+		errMessage  string
+	}{
+		{false, false, false, ""},
+		{true, false, true, fmt.Sprintf(expectedErrMessage, "c1")},
+		{false, true, true, fmt.Sprintf(expectedErrMessage, "c2")},
+		{true, true, true, fmt.Sprintf(expectedErrMessage, "c1")},
+	} {
+		c1 := getNoiselessCount()
+		c2 := getNoiselessCount()
+
+		if tc.callEncode1 {
+			c1.GobEncode()
+		}
+		if tc.callEncode2 {
+			c2.GobEncode()
+		}
+
+		err := checkMergeCount(c1, c2)
+		if (err != nil) != tc.wantErr {
+			t.Errorf("CheckMerge: No error was returned, expected error")
+		}
+		if tc.wantErr && err.Error() != tc.errMessage {
+			t.Errorf("CheckMerge: Error messages are not equal:\n got: %v\n want: %v", err.Error(), tc.errMessage)
 		}
 	}
 }
```

Comment:
We could probably fit these test as test cases in CheckMergeCompatibility. Instead of calling Merge() or GobEncode, we could just modify the state of the Count object.

Reply:
I am not sure how you can fit all these different tests into one function. Also, shouldn't we call the functions and let it modify the state of the object to check transitions rather than we change it instead?

Reply:
Merged them into CheckMergeStateChecks same with Serialization, This way I can use noiseless count instead of specifying count options on every test in CheckMergeCompatibility